rfc/0004-at_plus.md

```diff
@@ -0,0 +1,207 @@
+ - Current Status: Draft
+
+# Summary
+This RFC specifies the user API and interaction with the server for enhanced
+RYOW query performance (commonly known as `AT_PLUS` with Mutation Tokens).
+
+# Motivation
+The current approach to perform RYOW ("Read Your Own Writes") queries is to use the
+`REQUEST_PLUS` scan consistency when executing a N1QL query. While this is
+semantically correct, it takes the performance hit of having to pick the absolute
+latest sequence number in the system and as a result potentially waiting for
+a longer time until the indexer has caught up (especially under high mutation
+  rates).
+
+One way to mitigate these effects is to only wait until specific mutations have
+been indexed instead of waiting for all of them up to the latest one. Also, the
+likelihood of the indexer already having indexed the mutation is directly
+proportional to the time delta between mutation and subsequent query.
+
+Note that `AT_PLUS` is treated as a pure optimization over `REQUEST_PLUS` under
+certain workloads and has no semantic difference to it in terms of correctness
+inside the bounds specified through the mutation tokens. One cannot use
+`REQUEST_PLUS` and `AT_PLUS` together in the same query.
+
+# General Design
+In the big picture of `AT_PLUS` for the developer, there are two components that
+must be considered:
+
+ 1. Mutations are executed on a bucket. They might or might not include
+    `MutationTokens` as part of the response.
+ 2. The application state after the mutation happened needs to be applied at
+    N1QL query time.
+
+This RFC does not focus on 1. since it has been implemented already as part of
+the enhanced durability requirements initiative in the 4.0 timeframe. This RFC
+assumes that the Mutation Token is (at least) available as a return value from
+a mutation operation, most likely exposed on the `Document` or equivalent or 
+stored internally (see implementation details). `AT_PLUS` is not possible without
+`MutationTokens`.
+
+Note that this RFC proposes an extension of the current `MutationToken` to also
+include the bucket `name` as a reference. This is needed to prevent the user from
+mistakenly pass a mutation token from one bucket into another one (and to resolve
+issues around multi-bucket n1ql queries).
+
+Since it is important to this RFC, the properties of a `MutationToken` are as
+follows:
+
+ 1. A `MutationToken` is opaque to the user and needs to be treated as such.
+ 2. A `MutationToken` is immutable after creation.
+
+The internal structure of a Mutation Token is a triple which contains:
+
+```
+MutationToken {
+    [required] vbucket_id = long
+    [required] vbucket_uuid = long
+    [required] sequence_number = long
+    [required] bucket = string/reference (either name of the bucket or immutable "reference" to the target bucket)
+}
+```
+
+The N1QL query execution for AT_PLUS consists of the following user level API:
+
+```
+consistent_with(bucket: Bucket, [docs: Document...])
+```
```

Comment:
Ah, I guess we need to clarify in the spec, this is on the part where all the other query options are, associated with each query.

Actually I was thinking about this too, if no bucket is provided then the one in scope can be used. The only downside I see with this is that in a cluster-level setting it would need to throw a runtime exception saying it is ambiguous.